nxp: imx93_a55: Switch to timer domain and Zephyr native drivers

LaurentiuM1234 opened the pull request:
With this PR, the i.MX93 platform will now start using the timer domain in conjunction with the Zephyr native drivers.

Zephyr dependencies:

1. https://github.com/zephyrproject-rtos/zephyr/pull/65671 (EDMA)
2. https://github.com/zephyrproject-rtos/zephyr/pull/65187 (DAI)
3. https://github.com/zephyrproject-rtos/zephyr/pull/65679 (removal of SAI3 and EDMA static mappings)
4. https://github.com/zephyrproject-rtos/zephyr/pull/65684 (DAI_DIR_{TX/RX} value adjustment)

**Important notes**:

1. Currently, the PR is split in multiple commits. This is done to ease the reviewing process, but most of the commits will be squashed into a single one to avoid breaking git bisect.
2. If git dependency 3 is merged before this PR and the Zephyr revision is bumped up this will lead to SOF no longer working on i.MX93 because the SAI and EDMA address spaces will no longer have any mappings, leading to translation faults.
3. The transition to timer domain is mandatory because the DMA domain is unusable with the Zephyr native drivers ATM.

lgirdwood commented:
@LaurentiuM1234 I'm guessing someone from NXP will test  after Zephyr PRs merge and update topologies to use timer ?

LaurentiuM1234 commented:
> @LaurentiuM1234 I'm guessing someone from NXP will test after Zephyr PRs merge and update topologies to use timer ?

The Zephyr PRs will only affect i.MX93 which has one topology that's updated in this PR

lyakh commented on src/audio/host-zephyr.c:
```diff
@@ -213,6 +215,19 @@ static int host_copy_one_shot(struct host_data *hd, struct comp_dev *dev, copy_c
+		/* head block currently points to a stack address so use new
+		 * block for the re-configuration to avoid any "surprises".
+		 */
+		hd->z_config.head_block = &blk;
```
sorry, what exactly does your comment above mean? Indeed this code stores a stack address into a persistent object and that indeed isn't a good idea and should be fixed, so which new block is your comment referring to, that fixes any surprises?

LaurentiuM1234 replied:
Ah, what I wanted to say with that comment is that on each `host_copy_one_shot()` you use a new block placed on the stack (i.e: the `blk` variable).

LaurentiuM1234 replied:
Fixed. Hopefully it's a bit better now?

lyakh replied:
@LaurentiuM1234 but what my comment actually was around, is that you shouldn't store (at least not permanently) a pointer to an object on stack (`blk`) in a permanent object (`hd->z_config`). So that assignment isn't a very good idea. You can make it kind of acceptable if you reset the pointer to NULL again *inside* the same function before leaving it like
```
if (IS_ENABLED(CONFIG_DMA_NXP_SOF_HOST_DMA)) {
    ...
    hd->z_config.head_block = &blk;
    ...
}
...
ret = dma_config(hd->chan->dma->z_dev, hd->chan->index, &hd->z_config);
/* *** clear the pointer to a stack object! *** */
hd->z_config.head_block = NULL;
```
assuming that your `dma_config()` doesn't "steal" that pointer to store it in a different persistent object, but that would be a bug of its own then.

LaurentiuM1234 replied:
ACK. Then I think I'm going to go for allocating the block configuration on the heap just like we do on `dai-zephyr`.

LaurentiuM1234 replied:
dcd2bd31bf0de9cd698f3342440d8151abcaa1a8 should address this now

lyakh replied:
if that object has to be persistent, then yes, it has to be allocated on heap. But if it's only used by the function and from other functions, used from it, i.e. while the stack frame is valid, then you can keep it on stack and then reset the pointer to `NULL` at the end of the function.

lyakh commented on src/audio/host-zephyr.c:
```diff
@@ -842,11 +842,21 @@ int host_common_params(struct host_data *hd, struct comp_dev *dev,
 
+	dma_block_cfg = rballoc(SOF_MEM_FLAG_COHERENT,
+				SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_DMA,
+				sizeof(struct dma_block_config));
```
please use `rzalloc()` and remove `memset()` below

LaurentiuM1234 replied:
done

LaurentiuM1234 commented:
**V2 updates**

1. dma_block_cfg is now allocated on the heap.
2. `CONFIG_DAI=y` has been removed from i.MX93's configuration file. It's already enabled via `prf.conf`
3. Removed `CONFIG_SAI_HAS_XCR4_CHMOD` as a consequence to it being removed from the Zephyr driver.

keqiaozhang commented:
SOFCI TEST

LaurentiuM1234 commented:
**V3 updates**

1. Removed selection of `CONFIG_DMA_DOMAIN` from `CONFIG_IMX93_A55`. DMA domain is no longer used on i.MX93 so no need to compile `zephyr_dma_domain.c`.

lgirdwood commented:
@LaurentiuM1234 @dbaluta I've penciled in for v2.9.

dbaluta commented:
There is some issues with compilation on Windows:

https://github.com/thesofproject/sof/actions/runs/7555242327/job/20569791161?pr=8520

LaurentiuM1234 commented:
> There is some issues with compilation on Windows:
> 
> https://github.com/thesofproject/sof/actions/runs/7555242327/job/20569791161?pr=8520

That's normal ATM. Since we're still missing the EDMA driver, I can't bump up the Zephyr revision.

dbaluta commented:
Thanks! Now i see the DNM tag.

iuliana-prodan commented on app/boards/mimx93_evk_a55.overlay:
```diff
@@ -58,4 +46,47 @@
+	edma4: dma@42000000 {
+		compatible = "nxp,edma";
+		reg = <0x42000000 (DT_SIZE_K(64) * 32)>;
+		valid-channels = <0>, <1>;
```
If you have here `<0>, <1>` shouldn't be `dma-channels` based on this: https://github.com/zephyrproject-rtos/zephyr/pull/65671/commits/bba525577eb0e0528e1460cd2e43aa952958400d#diff-7e174a2aefc1b54220f6b815c04028a782f004bc4d5d03a39f604703f09b3234R16?

iuliana-prodan replied:
> Do you mean having something like: `dma-channels = <2>` instead of `valid-channels = <0>, <1>`? 

Yes.

>If so, both approaches are correct. For a (**very**) high number of channels, the `dma-channels` "way" would be better since the lookup function would be O(1), but since this will never be the case I'm not sure choosing one approach over the other would make that much difference. If you believe the `dma-channels` approach is more intuitive then I don't mind switching to it.

Ok, let's keep` valid-channels`, is not much difference for 2 channels. 
We'll change the list (`<0>, <1>`) when we'll have support for multiple channels and the update will be easier having already the `valid-channels` option.

iuliana-prodan commented on app/boards/mimx93_evk_a55.overlay:
```diff
@@ -58,4 +46,47 @@
+		 */
+		tx-fifo-watermark = <1>;
+		/* can't let SOF use the actual FIFO depth (128 FIFO words) because
+		 * the DMA transfer size isn't divizible by it so the driver would
```
s/divizible/divisible

iuliana-prodan commented on src/audio/host-zephyr.c:
```diff
@@ -878,12 +890,11 @@ int host_common_params(struct host_data *hd, struct comp_dev *dev,
 
 	memset(dma_cfg, 0, sizeof(*dma_cfg));
-	memset(&dma_block_cfg, 0, sizeof(dma_block_cfg));
 
```
You can do the allocation for `dma_block_cfg` here, and there's no need to add `rfree(dma_block_cfg)` above (line 858 and 877)

LaurentiuM1234 replied:
fixed, thanks!

LaurentiuM1234 commented:
**V5 updates**:

1. Most Zephyr dependencies have been merged, added west.yml bump. Waiting on https://github.com/zephyrproject-rtos/zephyr/pull/67907

LaurentiuM1234 commented:
**V6 updates**

1. Dependencies merged, updated Zephyr manifest and 64-bit CI build is now clean.
2. Switched to RX SYNC - TX ASYNC mode. See comments from overlay for explanation.

A few notes:

1. Current PR layout will break git bisect. Since 93 is not officially supported in the kernel and is mostly used to test out new stuff I would be inclined to keep the layout as-is as long as this won't affect anyone else.
2. Bunch of CI failures, some of them caused by the recent SMP changes from Zephyr.
3. Need to re-test everything to make sure the native switch works ok. Will lift the DNM tag once this is done and the SMP-related issues are fixed. Also need to make sure that this PR isn't a cause for the other non-SMP-related CI failures.

LaurentiuM1234 commented:
@lgirdwood @kv2019i do we, by any chance, have an ETA on the fix for the SMP-related issues? It would be very helpful. Thanks!

kv2019i commented:
We are very close to unblocking Zephyr updates now, but as there are still issues, we'll do SOF2.9 with old Zephyr baseline. I'll bump this to 2.10.

LaurentiuM1234 commented:
**V7 updates**

1. EDMA4 and SAI3 nodes have been moved to Zephyr (see https://github.com/zephyrproject-rtos/zephyr/pull/70045). All we do here is enable and configure them for SOF usage.
2. Removed commits that were already merged via https://github.com/thesofproject/sof/pull/8859
3. Removed cleanup commits. This will be handled separately.

LaurentiuM1234 commented:
**V8 updates**

1) All dependencies have been merged. Removing the [DNM] tag.
2) Added manifest update
